src/librustc_resolve/error_reporting.rs

```diff
@@ -0,0 +1,162 @@
+        mut path: Vec<Ident>
+    ) -> Option<Vec<Ident>> {
+        // Replace first ident with `self` and check if that is valid.
+        path[0].name = keywords::SelfValue.name();
+        let result = self.resolve_path(None, &path, None, false, span, CrateLint::No);
+        debug!("make_missing_self_suggestion: path={:?} result={:?}", path, result);
+        if let PathResult::Module(..) = result {
+            Some(path)
+        } else {
+            None
+        }
+    }
+
+    /// Suggest a missing `crate::` if that resolves to an correct module.
+    ///
+    /// ```
+    ///    |
+    /// LL | use foo::Bar;
+    ///    |     ^^^ Did you mean `crate::foo`?
+    /// ```
+    fn make_missing_crate_suggestion(
+        &mut self,
+        span: Span,
+        mut path: Vec<Ident>
+    ) -> Option<Vec<Ident>> {
+        // Replace first ident with `crate` and check if that is valid.
+        path[0].name = keywords::Crate.name();
+        let result = self.resolve_path(None, &path, None, false, span, CrateLint::No);
+        debug!("make_missing_crate_suggestion:  path={:?} result={:?}", path, result);
+        if let PathResult::Module(..) = result {
+            Some(path)
+        } else {
+            None
+        }
+    }
+
+    /// Suggest a missing `super::` if that resolves to an correct module.
+    ///
+    /// ```
+    ///    |
+    /// LL | use foo::Bar;
+    ///    |     ^^^ Did you mean `super::foo`?
+    /// ```
+    fn make_missing_super_suggestion(
+        &mut self,
+        span: Span,
+        mut path: Vec<Ident>
+    ) -> Option<Vec<Ident>> {
+        // Replace first ident with `crate` and check if that is valid.
+        path[0].name = keywords::Super.name();
+        let result = self.resolve_path(None, &path, None, false, span, CrateLint::No);
+        debug!("make_missing_super_suggestion:  path={:?} result={:?}", path, result);
+        if let PathResult::Module(..) = result {
+            Some(path)
+        } else {
+            None
+        }
+    }
+
+    /// Suggest a missing external crate name if that resolves to an correct module.
+    ///
+    /// ```
+    ///    |
+    /// LL | use foobar::Baz;
+    ///    |     ^^^ Did you mean `baz::foobar`?
+    /// ```
+    ///
+    /// Used when importing a submodule of an external crate but missing that crate's
+    /// name as the first part of path.
+    fn make_external_crate_suggestion(
+        &mut self,
+        span: Span,
+        mut path: Vec<Ident>
+    ) -> Option<Vec<Ident>> {
+        // Need to clone else we can't call `resolve_path` without a borrow error.
+        let external_crate_names = self.extern_prelude.clone();
+
+        // Insert a new path segment that we can replace.
+        let new_path_segment = path[0].clone();
+        path.insert(1, new_path_segment);
+
+        for name in &external_crate_names {
+            // Don't suggest meta as it will error in `resolve_path`.
+            if name.as_str() == "meta" {
```

Comment:
Hardcoding this is wrong - you want to use something like this instead: https://github.com/rust-lang/rust/blob/c222479c6feb036e5e5198e068aa9824d855ffa4/src/librustc_resolve/lib.rs#L4403-L4407